cirq/ops/two_qubit_diagonal_gate.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Creates the Gate instrance for a two qubit diagonal gate.
+
+The Gate is used to create a 4x4 matrix with the diagonal elements
+passed as a list.
+"""
+
+from typing import Any, Tuple, List, TYPE_CHECKING
+import numpy as np
+import sympy
+
+from cirq import protocols, value
+from cirq._compat import proper_repr
+from cirq.ops import gate_features
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    import cirq
+
+
+@value.value_equality()
+class TwoQubitDiagonalGate(gate_features.TwoQubitGate):
+    """A gate given by a diagonal 4x4 matrix."""
+
+    def __init__(self, diag_angles_radians: List[value.TParamVal]) -> None:
+        r"""A two qubit gate with only diagonal elements.
+
+        This gate's off-diagonal elements are zero and it's on diagonal
+        elements are all phases.
+
+        Args:
+            diag_angles_radians: The list of angles on the diagonal in radians.
+                If these values are $(x_0, x_1, \ldots , x_3)$ then the unitary
+                has diagonal values $(e^{i x_0}, e^{i x_1}, \ldots, e^{i x_3})$.
+        """
+        self._diag_angles_radians: Tuple[value.TParamVal, ...] = tuple(
+            diag_angles_radians)
+
+    def _is_parameterized_(self) -> bool:
+        return any(
+            isinstance(angle, sympy.Basic)
+            for angle in self._diag_angles_radians)
+
+    def _has_unitary_(self) -> bool:
+        return not self._is_parameterized_()
+
+    def _unitary_(self) -> np.ndarray:
+        if self._is_parameterized_():
+            return None
+        return np.diag(
+            [np.exp(1j * angle) for angle in self._diag_angles_radians])
+
+    def _apply_unitary_(self, args: 'protocols.ApplyUnitaryArgs') -> np.ndarray:
+        if self._is_parameterized_():
+            return NotImplemented
+        for index, angle in enumerate(self._diag_angles_radians):
+            subspace_index = args.subspace_index(big_endian_bits_int=index)
+            args.target_tensor[subspace_index] *= np.exp(1j * angle)
```

Comment:
What happens with https://github.com/quantumlib/Cirq/pull/2591/files#r355092574?

Reply:
`args.target_tensor *= np.exp(1j * self._diag_angles_radians.reshape((2,2)))` did not work since `self._diag_angles_radians` was not a `numpy.array`. Similar to above, the `np.exp(1j * self_diag_angles_radians)` didn't work with tuples.